### ISSUE TITLE
AWS Load Balancer Controller to register new pods first and then deregister old pods

### ISSUE THREAD

**Issue by dilipkraghupatruni** [kubernetes-sigs/aws-load-balancer-controller#3621]
**Is your feature request related to a problem?**
We are using Argo rollouts for application deployments to EKS. 
- Argo rollouts performs a blue-green flip by updating the selectorLabels on a Service object.
- When "promote" is triggered, Argo switches the rollout template hash on active service to use the pod-rollout-template-hash from the new set (rollout)
- When this happens, the aws-load-balancer-controller does a complete swap of all members of the target group.

During blue-green switch, we are noticing issues related to AWS Load Balancer Controller when used with NLB. NLB takes lot more time to register new targets compared to ALB. Because of the current logic in aws load balancer controller, the old pods are deregistered first and then load balancer controller tries to register the new pods. However, the lag between these two steps is significant enough with NLB to cause app impact. The order of the operation is same with ALB but ALB does the operation very fast compared to NLB and hence we dont notice any impact. 

**Describe the solution you'd like**
We would like to get a feature toggle/ config in load balancer controller to swap the operations which means we want to register new pods first and then deregister old pods. https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/main/pkg/targetgroupbinding/resource_manager.go#L143-L152
And also configure the amount of time we want to wait between the operations and also max amount of time we want to wait on deregistration.

**Describe alternatives you've considered**
We have implemented canary deployment strategy instead of blue green and that helped us reduce the impact. But as part of the debugging, we figured out the problem is not specific to NLB. We deal with thousands of applications where majority use blue-green deployment strategy. hence we cannot implement canary for everyone. We would like to fix the problem at its core which is inside aws load balancer controller.

**Comment by jwenz723** [comment]
> NLB takes lot more time to register new targets compared to ALB. Because of the current logic in aws load balancer controller, the old pods are deregistered first and then load balancer controller tries to register the new pods. However, the lag between these two steps is significant enough with NLB to cause app impact. The order of the operation is same with ALB but ALB does the operation very fast compared to NLB and hence we dont notice any impact.

Just wanted to clarify that the amount of time between when [deregistration executes](https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/83931921cca2e61f2c719226116f80e86219401c/pkg/targetgroupbinding/resource_manager.go#L144) and when [registration executes](https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/83931921cca2e61f2c719226116f80e86219401c/pkg/targetgroupbinding/resource_manager.go#L149) is negligible. The deregistration and registration events are essentially occurring at the same time (probably just a few milliseconds between these events).

The problem is that the old targets/pods enter a `deregistering` state in the NLB Target Group at the same time as when the new targets/pods enter an `initial` state in the NLB Target Group and NLB Target Group registration takes between 3 to 5 minutes (see [this comment](https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/1834#issuecomment-1728188047)). This means that for 3 to 5 minutes there are 0 `healthy` targets in the NLB Target Group.

Regarding ALBs, I've observed during my testing that ALB Target Group registration typically takes between 10-15 seconds. So during a blue-green deploy with an ALB there is a 10-15 second period of time when there are 0 `healthy` targets in the ALB Target Group.

Having a quantity of time to wait after target registration and before target deregistration as mentioned by @dilipkraghupatruni  should help to maintain `healthy` targets in the NLB/ALB Target Group during a blue-green flip.

**Comment by M00nF1sh** [comment]
The problem here is it's hard for the controller to decide whether a old target(that no longer is a valid target) shall be removed. In your case, the target shall be kept for 2~4 mins(for NLB), while in most other cases, such targets shall be deregistered asap(as they may belong to other applications or already died).

**Comment by millermatt** [comment]
> The problem here is it's hard for the controller to decide whether a old target(that no longer is a valid target) shall be removed. In your case, the target shall be kept for 2~4 mins(for NLB), while in most other cases, such targets shall be deregistered asap(as they may belong to other applications or already died).

@M00nF1sh : part of @dilipkraghupatruni 's suggestion was to add config to control this behavior. That would take the decision making out of the controller's hands. Rather, it would do the operations in the order specified in the config and at the intervals specified in the config.

**Comment by vl-kp** [comment]
any update on this? targets in initial or deregistration phase won't receive new connections

**Comment by youwalther65** [comment]
This is an interesting problem I haven't thought about before. Usually users use deployments which handle rollouts of new pods according to deployment's `spec.strategy` for example using `maxSurge` and `maxUnavailable`, so there is always a registered and healthy pod to serve traffic. Is there no way of Argo using this approach?
Regarding time it takes for NLB to register a target this has improved a lot and we see NLB targets going healthy in about 90s now (including initial health check with 3 times 10s health check), see [here](https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/1834#issuecomment-2359354913)

**Comment by youwalther65** [comment]
Thank you for sharing this @jwenz723.
So we at least have a solution for ALB fronted services. Not yet for NLB fronted services because NLB does not support weighted target groups.

**Comment by yossisolomon** [comment]
Hey, we're handling a similar issue but with an ALB - we had 5 seconds of downtime during one of our recent switches, so not minutes but still too impactful for us to just not handle it (in our case it's possible it will have impact on users as not all requests are retryable, so we're trying to avoid it as much as possible). This made us stop an ongoing migration of one of our biggest user-facing services to K8s.

I'm not sure why this wasn't mentioned before like the PingPong feature was, but there is an[ ArgoRollouts flag that seems to help with this situation](https://argo-rollouts.readthedocs.io/en/stable/features/traffic-management/alb/#zero-downtime-updates-with-aws-targetgroup-verification) - `--aws-verify-target-group` which is supposed to verify the switch is handled well at the TargetGroup level.

However, the issue with that is the verification only happens _**after**_ moving the service reference to the new ReplicaSet - so deregistration will still happen first and it will take some random amount of seconds for the ALB first Healthchecks to pass and make the new targets turn from Initial to Healthy state, which is still too late as in the interim we have 0 healthy targets and will get 503s (Service Unavailable) from the ALB.

We are currently considering orchestrating this manually using the underlying spec objects and AWS TG APIs to verify the rollout is safe - of course we would be happy to avoid this, but don't see a clear solution that properly handled this situation for blue-green.

For PingPong - we simply don't understand from the docs what actually happens under the hood to understand if it can work for us. E.g. we don't know how we control each ping/pong service (we need production traffic on one and our orchestrated test traffic to go to the other as done in blue-green with the previewService), how those interplay with the ReplicaSets and the TargetGroupBinding or the LB healthchecks - so other than experimenting with it (which could be a waste of time) we don't see a clear path to understand how this works. This is more of an issue for the ArgoRollouts repo TBH, but wanted to share this context here too.

It would have been great to be able to just control the phasing of the registration+deregistration so that we can control it's ordering to avoid 503s (we get why this can be complex and are ok with this being via overrides or something else we can automate via the K8s API).

**Comment by jwenz723** [comment]
@yossisolomon the aws-load-balancer-controller is by design setup to sync the pods which are mapped to a kubernetes service (endpoint slice) into an AWS target group exactly as they exist in the endpoint slice. If it is desired to add a delay between registration and deregistration to ensure the 'green' pods are healthy in the ALB before the 'blue' pods are deregistered then I think this modification should be made in the Argo Rollouts controller not in the aws-load-balancer-controller because the Argo Rollouts controller is controlling which pods are mapped to the kubernetes service.

# Why BlueGreen doesn't work with ALB  'ip' target type

The Argo Rollouts BlueGreen deployment strategy by default executes the cutover from the 'blue' pods to the 'green' pods by modifying the labelSelector on a kubernetes service. Before the cutover, the labelSelector on the service is configured to point at the 'blue' pods. After the cutover (i.e. promote) the labelSelector is configured to point at the 'green' pods. This action of modifying the labelSelector on the kubernetes service causes all of the 'blue' pods to almost instantly be removed from the service (endpoint slice) and causes all of the 'green' pods to be almost instantly added into the service (endpoint slice). The aws-load-balancer-controller observes that all 'blue' pods were removed from the service and all 'green' pods were added into the service and this triggers the aws-load-balancer-controller to deregister all 'blue' pods and register all 'green' pods in the associated AWS target group so that the target group members match the kubernetes service members.

This is OK when you use 'instance' type routing on your ALB because the traffic coming into your ALB flows from the ALB to a kubernetes node and then from the node to one of your application pods. That routing from the node to the pod is performed according to the pods which are included in the kubernetes service (endpoint slice), which is updated almost instantly.

This is NOT OK when you use 'ip' type routing on your ALB because the traffic coming into your ALB flows from the ALB directly to your application pods. Registration and deregistration of pods into your ALB target groups is not instant like the kubernetes service (endpoint slice) update. This non-instantaneous update to the target groups is what results in your observed 5 seconds of getting 503s from the ALB.

The ping-pong deployment strategy works well and I believe is the solution to this problem. I agree, as you stated @yossisolomon, that you should test out the strategy to ensure it works as desired for your use case before rolling the change into any critical environments.  

# How ping-pong works with a BlueGreen step configuration

The pingPong feature is only supported when using the `canary` strategy. However, as I mentioned in [my post above](https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/3621#issuecomment-2484801962) you can use a step configuration to recreate a BlueGreen style deployment using this strategy using steps like this:

```yaml
  strategy:
    canary:
      pingPong:
        pingService: ping-svc
        pongService: pong-svc
      steps:
        - setWeight: 0
        - setCanaryScale:
            weight: 100
        - pause: {}
      trafficRouting:
        alb:
          ingress: active-ingress
          rootService: root-svc
          servicePort: 443
```

With the pingPong strategy configured with the code above, the Argo Rollouts controller injects the `alb.ingress.kubernetes.io/actions.<app name>-root-svc` annotation into your Kubernetes Ingress with a value that contains 2 target groups (i.e. a ping target group and a pong target group). Both the ping and pong target groups are considered as stable/active target groups, however, only one of them will receive traffic at any point in time. This is because the 2 target groups are added into the ALB with a traffic weighting percentage which specifies how much of the incoming traffic should be routed to each of the ping and pong target groups. You can read more about ALB traffic weighting [here](https://aws.amazon.com/blogs/aws/new-application-load-balancer-simplifies-deployment-with-weighted-target-groups/). With the configuration mentioned above, one of the ping and pong target groups with be configured to receive 0% of the incoming traffic and the other will be configured to receive 100% of the incoming traffic. You can optionally add more steps into your configuration to cause the switch to be more gradual if you want a canary style gradual cutover.

When you perform a deployment of a new version of your application the Argo Rollouts controller will designate the target group which is currently receiving 0% of traffic as the new stable/active target group. The Argo Rollouts controller will provision all of your new pods and the new stable/active kubernetes service will be updated to point to these new pods. The update to that kubernetes service will be observed by the aws-load-balancer-controller, which will result in the new pods being registered into the associated AWS target group. After the targets in the target group are considered healthy then the 'promote' operation on the Argo Rollout can be invoked to cause the Argo Rollouts controller to update the `alb.ingress.kubernetes.io/actions.<app name>-root-svc` annotation on your Kubernetes Ingress to have 100% of traffic going to the new stable/active target group and 0% of traffic going to the old stable/active target group. This annotation modification will be observed by the aws-load-balancer-controller, which will perform an AWS API call to update the traffic weighting of the target groups accordingly. This AWS API call is able to cutover the traffic from the old target group to the new target group without creating or destroying any target groups. The only thing that occurs is the traffic weighting percentage is modified. This means that the update can take effect immediately (i.e. no more 5 second period of time where 503s are returned by the ALB).

Hope this helps.

**Comment by yossisolomon** [comment]
Thanks for the detailed reply @jwenz723 ! 
Unfortunately we have a similar situation to what is mentioned [here](https://github.com/argoproj/argo-rollouts/issues/4433) where we'd like the preview setup to be stable (we're running test traffic through it to test out the new version) and moving weights and TG bindings around on the prod endpoint each time we do a release feels really error prone.

We did end up setting up our own blue-green orchestrated setup that validates the registration, but we had to dig deep into Argo's code to get it to work properly. So we'll keep trying to push for the resolution on their end. Hopefully your comment about that will help.

**Comment by mikebordon** [comment]
The ping-pong solution is a good workaround for Ingress implementations, but that rollout strategy is specific to ALB and Istio routers only, so it is not an option for Gateway API (HTTPRoute) implementations which ultimately suffer from this same deregistration+registration blip during rollout completion due to the service selector swap.

I don't think the burden should necessarily be on the Rollouts devs to port the ping-pong strategy to Gateway API since it's (mostly) solving an ALB-specific problem and Gateway API is intentionally provider-agnostic.

Or maybe this warrants a separate/new issue specific to the Gateway API implementation?
